src/interfaces/cloud.rs

```diff
@@ -0,0 +1,410 @@
+    pub fn random_findex_master_key(
+        index_id: String,
+        fetch_entries_key: KeyingMaterial<SIGNATURE_KEY_LENGTH>,
+        fetch_chains_key: KeyingMaterial<SIGNATURE_KEY_LENGTH>,
+        upsert_entries_key: KeyingMaterial<SIGNATURE_KEY_LENGTH>,
+        insert_chains_key: KeyingMaterial<SIGNATURE_KEY_LENGTH>,
+    ) -> Result<Self, FindexErr> {
+        let mut rng = CsRng::from_entropy();
+        let findex_master_key = KeyingMaterial::<MASTER_KEY_LENGTH>::new(&mut rng);
+
+        Ok(Token {
+            index_id,
+            findex_master_key,
+            fetch_entries_key: Some(fetch_entries_key),
+            fetch_chains_key: Some(fetch_chains_key),
+            upsert_entries_key: Some(upsert_entries_key),
+            insert_chains_key: Some(insert_chains_key),
+        })
+    }
+
+    pub fn reduce_permissions(&mut self, search: bool, index: bool) -> Result<(), FindexErr> {
+        self.fetch_entries_key =
+            reduce_option("fetch entries", &self.fetch_entries_key, search || index)?;
+        self.fetch_chains_key = reduce_option("fetch chains", &self.fetch_chains_key, search)?;
+        self.upsert_entries_key = reduce_option("upsert entries", &self.upsert_entries_key, index)?;
+        self.insert_chains_key = reduce_option("insert chains", &self.insert_chains_key, index)?;
+
+        Ok(())
+    }
+}
+
+/// If we have the permission and want to keep it, do nothing.
+/// If we don't have the permission and want to keep it, fail.
+/// If we don't want to keep it, return none.
+fn reduce_option(
+    debug_info: &str,
+    permission: &Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    keep: bool,
+) -> Result<Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>, FindexErr> {
+    match (permission, keep) {
+        (_, false) => Ok(None),
+
+        (Some(permission), true) => Ok(Some(permission.clone())),
+        (None, true) => Err(FindexErr::Other(format!(
+            "The token provided doesn't have the permission to {debug_info}"
+        ))),
+    }
+}
+
+enum Callback {
+    FetchEntries,
+    FetchChains,
+    UpsertEntries,
+    InsertChains,
+}
+
+impl Callback {
+    pub fn get_uri(&self) -> &'static str {
+        match self {
+            Callback::FetchEntries => "fetch_entries",
+            Callback::FetchChains => "fetch_chains",
+            Callback::UpsertEntries => "upsert_entries",
+            Callback::InsertChains => "insert_chains",
+        }
+    }
+
+    pub fn get_key<'a>(
+        &self,
+        token: &'a Token,
+    ) -> &'a Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>> {
+        match self {
+            Callback::FetchEntries => &token.fetch_entries_key,
+            Callback::FetchChains => &token.fetch_chains_key,
+            Callback::UpsertEntries => &token.upsert_entries_key,
+            Callback::InsertChains => &token.insert_chains_key,
+        }
+    }
+}
+
+impl FindexCloud {
+    pub fn new(token: String, base_url: Option<String>) -> Result<Self, FindexErr> {
```

Comment:
You don't need the ownership here.
```suggestion
    pub fn new(token: &str, base_url: Option<String>) -> Result<Self, FindexErr> {
```